File: 백준/Silver/2841. 외계인의 기타 연주/외계인의 기타 연주.py
# https://www.acmicpc.net/problem/2841
from collections import defaultdict
n, p = map(int, input().split())
# 각 줄은 독립적!
guitar_strings_melody = defaultdict(list)

for _ in range(n):
    string_number, plat_number = map(int, input().split())
    guitar_strings_melody[string_number].append(plat_number)
    
def count_finger_number(plats):
    stack = []
    cnt = 0

    for plat in plats:
        # 새로운 plat이 stack에 마지막이 될 때까지 pop
        while stack and stack[-1] > plat:
            stack.pop()
            cnt += 1
        # stack이 비어있거나 새로운 plat이 가장 큰 경우        
        if not stack or stack[-1] < plat:
            stack.append(plat)
            cnt += 1

    return cnt

ans = 0
for key in guitar_strings_melody.keys():
    ans += count_finger_number(guitar_strings_melody[key])
print(ans)

File: 백준/Silver/2841. 외계인의 기타 연주/외계인의 기타 연주.java
import java.io.*;
import java.util.*;

public class Main {

    public static int n, m; 

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer inputNM = new StringTokenizer(br.readLine());

        n = Integer.parseInt(inputNM.nextToken());
        m = Integer.parseInt(inputNM.nextToken());

        Map<Integer, List<Integer>> guitarMelody = new HashMap<>();

        for (int i = 0; i < n; i++) {
            StringTokenizer row = new StringTokenizer(br.readLine());

            int stringNumber = Integer.parseInt(row.nextToken());
            int platNumber = Integer.parseInt(row.nextToken());

            List<Integer> platList = guitarMelody.get(stringNumber);
            if (platList == null) {
                platList = new ArrayList<>();
                guitarMelody.put(stringNumber, platList);
            }
            platList.add(platNumber);
        }

        int ans = 0;
        for (List<Integer> plats : guitarMelody.values()) {
            ans += countFingerNumber(plats);
        }

        System.out.println(ans);

    }

    public static int countFingerNumber(List<Integer> plats) {
        Deque<Integer> stack = new ArrayDeque<>();  
        int cnt = 0;

        for (int plat : plats) {
            while (!stack.isEmpty() && stack.peek() > plat) {
                stack.pop();
                cnt++;
            }
            if (stack.isEmpty() || stack.peek() < plat) {
                stack.push(plat);
                cnt++;
            }
        }
        return cnt;
    }
}